### PR TITLE
Move packages 'python3-*' & 'virtualenv' (for Python 2) from scripts/ansible to roles that need them, e.g. {mysql, postgresql, munin, kalite}

### PR DESCRIPTION
**Pull request by holta**
Per https://github.com/iiab/iiab/pull/2879#issuecomment-888327576

@jvonau might have further suggestions?

- PR #2527 
- PR #2655 
- PR #2659
- PR #2743

### CODE REVIEW

**Comment by holta** [comment]
@jvonau what OS is most useful to test here?

(Debian 11 presumably ?)

**Comment by holta** [comment]
I did an install test on Debian 11, stripping out 7 of the 8 apt packages from scripts/ansible (preserving only 'python3-pip', which is necessary to install ansible-core).

Which in turn dragged in 'python3-setuptools'.

The test was successful, so installation of package 'virtualenv' will move to kalite/tasks/install.yml

Which suggests that the remaining 5 packages (below) need further investigation, as to which roles truly do and do not need them:

- python3-pymysql (which MySQL-related roles fail without this?)
- python3-psycopg2 (which PostgreSQL-related roles fail without this?)
- python3-passlib
- python3-packaging
- python3-venv

Summary Question: which IIAB roles fail when the 5 packages above are not installed?

**Comment by holta** [comment]
I'll try to work on this overnight if I have time.

I don't have automated testing facilities to run all ~30 IIAB apps standalone (without any other apps) to "keep everyone honest."

(But hopefully together we can get much closer to the truth, allowing lightweight IIAB's to carry around much less deadweight.)

**Comment by holta** [comment]
Observations:

1) In short, this PR could be merged.  But I'd rather tighten up [scripts/ansible](https://github.com/iiab/iiab/blob/master/scripts/ansible) first &mdash; so it "does what it says, and says what it does."  i.e. installs Ansible and not much else.

2) Which doesn't mean this python3-* document should go away: https://github.com/iiab/iiab/blob/master/scripts/ansible.md
But hopefully this PR (or similar) will allow the above doc to be more informative 2 years later, with a more modern context.

3) Just FYI 'python3-pymysql' drags in 'python3' which drags in 'python3-venv' on Debian 11 here.

**Comment by holta** [comment]
> 3\. Just FYI `python3-pymysql` drags in `python3` which drags in `python3-venv` on Debian 11 here.

In short, installing `python3-pymsql` within roles/mysql (mandatory at the moment) is currently a very sloppy way to force the installation of `python3-venv` across the board.

(It would be far better to identify which IIAB roles truly need/use Python 3 venv's.)

**Comment by holta** [comment]
BIG-sized IIAB installs onto Debian 11 VM with this revised PR as is!

But there might be room for further improvement before merging.

(e.g. cleaning up or deprecating document https://github.com/iiab/iiab/blob/master/scripts/ansible.md at the very minimum!)

**Comment by jvonau** [comment]
Now what split out the collections? or at least community.postgresql to the on demand role? The others seem part of the overall stock install so that is why those apt packages were installed at that point was to support the 'collections'.

**Comment by jvonau** [comment]
> > 3. Just FYI `python3-pymysql` drags in `python3` which drags in `python3-venv` on Debian 11 here.
> 

Correct, python3-venv is a virtual subpackage of python3 it's a builtin module in newer python3 now... you get the '-m(odule) venv' automatically as used below

> In short, installing `python3-pymsqi` within roles/mysql (mandatory at the moment) is currently a very sloppy way to force the installation of `python3-venv` across the board.
> 

It's a builtin module in 'python3' and depends on python3-minimal which I would think lacks the full suite of modules.
> (It would be far better to identify which IIAB roles truly need/use Python 3 venv's.) 

ubuntu@box:~$ grep -r virtualenv /opt/iiab/iiab/roles/
```
/opt/iiab/iiab/roles/calibre-web/tasks/install.yml:    virtualenv: "{{ calibreweb_venv_path }}"    # /usr/local/calibre-web-py3
/opt/iiab/iiab/roles/calibre-web/tasks/install.yml:    virtualenv_site_packages: no
/opt/iiab/iiab/roles/calibre-web/tasks/install.yml:    virtualenv_command: python3 -m venv {{ calibreweb_venv_path }}
/opt/iiab/iiab/roles/lokole/tasks/install.yml:    virtualenv: "{{ lokole_venv }}"
/opt/iiab/iiab/roles/lokole/tasks/install.yml:    virtualenv_command: python3 -m venv "{{ lokole_venv }}"
/opt/iiab/iiab/roles/lokole/tasks/install.yml:    virtualenv: "{{ lokole_venv }}"
/opt/iiab/iiab/roles/lokole/tasks/install.yml:    virtualenv_command: python3 -m venv "{{ lokole_venv }}"
/opt/iiab/iiab/roles/lokole/tasks/install.yml:    virtualenv: "{{ lokole_venv }}"
/opt/iiab/iiab/roles/lokole/tasks/install.yml:    virtualenv_command: python3 -m venv "{{ lokole_venv }}"
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv: "{{ kalite_venv }}"    # /usr/local/kalite/venv
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv_site_packages: no
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv_command: /usr/bin/virtualenv
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv_python: python2.7
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv: "{{ kalite_venv }}"    # /usr/local/kalite/venv
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv_site_packages: no
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv_command: /usr/bin/virtualenv
/opt/iiab/iiab/roles/kalite/tasks/install.yml:    virtualenv_python: python2.7
/opt/iiab/iiab/roles/kalite/README.rst:*In late 2017, Internet-in-a-Box added a virtual environment (/usr/local/kalite/venv/) to keep KA Lite's Python package/dependency risks under control.  As such the command* `/usr/bin/kalite <https://github.com/iiab/iiab/blob/master/roles/kalite/templates/kalite.sh.j2>`_ *is a wrapper to this virtualenv*.  **Consequently, the following steps are no longer needed:**
/opt/iiab/iiab/roles/jupyterhub/tasks/install.yml:    virtualenv: "{{ jupyterhub_venv }}"    # /opt/iiab/jupyterhub
/opt/iiab/iiab/roles/jupyterhub/tasks/install.yml:    virtualenv_site_packages: no
/opt/iiab/iiab/roles/jupyterhub/tasks/install.yml:    virtualenv_command: /usr/bin/virtualenv
/opt/iiab/iiab/roles/jupyterhub/tasks/install.yml:    virtualenv_python: python3
/opt/iiab/iiab/roles/jupyterhub/templates/jupyterhub_config.py:#  python environment (with virtualenv/conda) than JupyterHub itself.
/opt/iiab/iiab/roles/jupyterhub/templates/jupyterhub_config.py:#  environment, such as when they have their own conda environments / virtualenvs
```

**Comment by holta** [comment]
> Now what split out the collections? or at least community.postgresql to the on demand role?

It's an idea.  Conversely it might turn out to be a lot more trouble than it's worth to stop and restart Ansible (or retrofit some other conditional arrangement) to economize 2 MB here:

```
root@box:~# du -hsc /usr/share/ansible/collections/ansible_collections/*/*
1.7M    /usr/share/ansible/collections/ansible_collections/ansible/posix
20M     /usr/share/ansible/collections/ansible_collections/community/general
1.1M    /usr/share/ansible/collections/ansible_collections/community/mysql
2.2M    /usr/share/ansible/collections/ansible_collections/community/postgresql
25M     total
```

- #2659

**Comment by holta** [comment]
> ubuntu@box:~$ grep -r virtualenv /opt/iiab/iiab/roles/

In summary, IIAB has 3 roles that use a Python 3 venv:

- calibre-web [Stage 9]
- jupyterhub [Stage 6 — this wasn't completely transitioned from Python 2 to Python 3, but this PR should fix that]
- lokole [Stage 6]

And 1 role that uses a Python 2 virtualenv:

- kalite [Stage 7]

**Comment by holta** [comment]
FYI it turns out IIAB has fewer roles using MySQL than I imagined.  The total includes 3 "community" roles:

- mediawiki
- nextcloud
- wordpress

And 2 "lesser-used" roles:

- elgg
- pbx (FreePBX)

FWIW all 5 are considered "generic" i.e. every one is a part of Stage [6-generic-apps](https://github.com/iiab/iiab/tree/master/roles/6-generic-apps).

[PS Admin Console's [js-menu](https://github.com/iiab/iiab-admin-console/tree/master/roles/js-menu) also uses MySQL]

**Comment by jvonau** [comment]
> > ubuntu@box:~$ grep -r virtualenv /opt/iiab/iiab/roles/
> 
> In summary, IIAB has 3 roles that use a Python 3 venv:
> 
>     * calibre-web
> 
>     * jupyterhub
> 
This one is using `virtualenv_command: /usr/bin/virtualenv` so it needs the same treatment as kalite or move to use `python3 -m venv` like the other two  

>     * lokole
> 
> 
> And 1 role that uses a Python 2 virtualenv:
> 
>     * kalite

**Comment by jvonau** [comment]
My suggestion is to drop this PR altogether, there is no gain by moving the supporting packages for the collections and why the packages for the dependencies are installed at the location was fully documented in https://github.com/iiab/iiab/blob/master/scripts/ansible.md when I wrote it 2 years ago as an authoritative source not the wishy washy dribble that questions itself `for Good Measure?` that it has become.

**Comment by tim-moody** [comment]
This sort of rearchitecting belongs at the beginning, not the end, of a release cycle.
Who is the target audience for these changes?

**Comment by jvonau** [comment]
Just needless churn as a feeble attempt to appear as the main authority.

File: roles/postgresql/tasks/install.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@
     name:
       - postgresql
       - postgresql-client
+      - python3-psycopg2    # Moodle install fails without this
```
Comment (jvonau) [inline roles/postgresql/tasks/install.yml]: To support Ansible modules: [postgresql_db](https://docs.ansible.com/ansible/latest/modules/postgresql_db_module.html) and [postgresql_dbuser](https://docs.ansible.com/ansible/latest/modules/postgresql_user_module.html) available from community.postgresql

File: roles/mysql/tasks/install.yml [hunk 2]
```diff
@@ -12,6 +12,7 @@
       - mariadb-client
       #- php{{ php_version }}-common    # Auto-installed as an apt dependency.  REGARDLESS: php{{ php_version }}-common superset php{{ php_version }}-cli is auto-installed by php{{ php_version }}-fpm in nginx/tasks/install.yml
       - php{{ php_version }}-mysql      # Likewise installed in nextcloud/tasks/install.yml, pbx/tasks/freepbx_dependencies.yml, wordpress/tasks/install.yml
+      - python3-pymysql                 # Required by MySQL roles like Elgg, MediaWiki etc
```
Comment (jvonau) [inline roles/mysql/tasks/install.yml]: Ansible modules: [mysql_db](https://docs.ansible.com/ansible/latest/modules/mysql_db_module.html) and [mysql_user](https://docs.ansible.com/ansible/latest/modules/mysql_user_module.html) available from community.mysql

File: roles/munin/tasks/install.yml [hunk 3]
```diff
@@ -1,11 +1,12 @@
-- name: "Install 4 packages: munin, munin-node, munin-plugins-extra, libcgi-fast-perl"
+- name: "Install 5 packages: munin, munin-node, munin-plugins-extra, libcgi-fast-perl, python3-passlib"
   package:
     name:
       - munin
       - munin-node
       - munin-plugins-extra
       - libcgi-fast-perl
       #- libapache2-mod-fcgid
+      - python3-passlib
```
Comment (jvonau) [inline roles/munin/tasks/install.yml]: Ansible module: [htpasswd](https://docs.ansible.com/ansible/latest/modules/htpasswd_module.html)

File: roles/2-common/tasks/packages.yml [hunk 4]
```diff
@@ -30,6 +30,8 @@
       #- openssh-server    # ssh (Raspbian) or openssh-server (other OS's) already installed by 1-prep's roles/sshd/tasks/main.yml
       - pandoc             # For /usr/bin/iiab-refresh-wiki-docs
       - pastebinit         # For /usr/bin/iiab-diagnostics
+      #- python3-pip       # Already installed by /opt/iiab/iiab/scripts/ansible -- this auto-installs 'python3-setuptools' and 'python3' etc
+      - python3-venv       # 2021-07-29: Unlike Debian 11, this is required on Ubuntu 20.04 (where packages like python3-pymysql -> python3 don't drag in 'python3-venv') -- for roles like {calibre-web, jupyterhub, lokole} -- whereas roles/kalite uses (virtual) package 'virtualenv' for Python 2
```
Comment (jvonau) [inline roles/2-common/tasks/packages.yml]: For use with Ansible module: [pip](https://docs.ansible.com/ansible/latest/modules/pip_module.html) -m venv

File: roles/kalite/tasks/install.yml [hunk 5]
```diff
@@ -15,7 +15,8 @@
   package:
     name:
       - python2
-      - python-setuptools # provides setuptools-44 last version compatible with python2
+      - python-setuptools    # Provides setuptools-44 (last version compatible with python2)
+      - virtualenv           # Provides venv for python2 -- compare python3-venv in scripts/ansible
```
Comment (jvonau) [inline roles/kalite/tasks/install.yml]: For use with Ansible module: [pip](https://docs.ansible.com/ansible/latest/modules/pip_module.html) virtualenv_command: /usr/bin/virtualenv virtualenv_python: python2.7

File: roles/postgresql/tasks/install.yml (hunk 1 above)
Comment (holta) [reply roles/postgresql/tasks/install.yml]: Correct spelling is `postgresql_user`

Thanks as this documentation really helps.

**Comment by jvonau** [comment]
All culled from ansible.md.unused... Just walking away now.

**Comment by holta** [comment]
For communities wanting lightweight IIAB's with far less bloat, or a focus on content with "no apps at all."

Tested intensively across of dozens OS/hardware combinations over 8 days now, with @lemueldsouza's assistance:

- PR #2891
- PR #2893
- PR #2896 
- PR #2899
- PR #2903
- PR #2906

**Pull request merged**